src/nvim/search.c

```diff
@@ -3878,24 +3881,34 @@ current_search (
 
 /*
  * Check if the pattern is one character or zero-width.
+ * If move is true, check from the beginning of the buffer,
+ * else from the current cursor position.
  * Returns TRUE, FALSE or -1 for failure.
  */
-static int is_one_char(char_u *pattern)
+static int is_one_char(char_u *pattern, bool move)
 {
   regmmatch_T regmatch;
   int nmatched = 0;
   int result = -1;
   pos_T pos;
   int save_called_emsg = called_emsg;
+  int flag = 0;
 
   if (search_regcomp(pattern, RE_SEARCH, RE_SEARCH,
           SEARCH_KEEP, &regmatch) == FAIL)
     return -1;
 
   /* move to match */
-  clearpos(&pos);
+  if (move) {
+      clearpos(&pos);
+  }
+  else {
```

Comment:
Sorry to be so nit-picky, but I just noticed that the indentation seems to be 4 spaces. It should be two spaces according to the [style guide](http://neovim.org/develop/style-guide.xml). Also, while I can't actually find it in the style guide right now, you can also move up the else (like `} else {`). This is what's used [elsewhere](https://github.com/neovim/neovim/commit/2e873c8767056e75e8e44d878caf5673df06f423#diff-a6109b430c5b2379023dc39efac014dcR330) in the codebase.